### ISSUE TITLE
ECE ALKS driver model inside esmini as controller (internal or external) or xosc (separated catalogue or not)

### ISSUE THREAD

**Issue by tmIdiada** [esmini/esmini#160]
Hi,

I try to integrate the ECE ALKS reference driver into esmini.
Now I have different possibilities to do that:

1. Define it as a sequence of actions into a separated catalogue (better solution for others) or directly into the xosc files. Where it would be defined plays for the functionality no role and also not for me, because all xosc files are generated automatically via scenariogeneration and must only work and not look fine). But I have to define a ramp up in deceleration from 0G until 0.774G within 6sec. Is it possible to define such a ramp for acceleration which would be executed by esmini exactly in that linear way?
2. The even better solution (what I'm trying actually) integrate it via controller (external or in esmini integrated). I've copied simply the ACC controller from esmini and did the corresponding modifications, that the new controller is found. I think the new controller would work, but there is one functionality missing: After a specific event happened, like a specific distance smaller a value, then the controller shall wait for 0.75sec and afterwards it shall execute the above ramp up in deceleration. The ramp up is easily to implement in the controller, but the wait time, this is my biggest issue. My current idea is to add a property to this controller class waittime and as long as timestep is lower than the waittime do nothing and set new waittime equal old one minus timestep, I think this way it would work, do you have a better solution in mind? And I have to be prepared that this contoller will be not merged into esmini, how one can define here external controllers?

One additional question: I need the lateral distance of 2 entities and I need the lateral distance to its lane center, I think the first one would be the difference of pos_.GetT(), right? But the second one? Where is the t of the center of the laneID stored?

Kind regards

Thaddäus

**Comment by eknabevcc** [comment]
1. Unfortunately there is yet no easy way of defining ramped acceleration, i.e. jerk. This is something that will be considered for OpenSCENARIO 1.2. So actually I take your question as another user request for such a feature :)
The only approximate solution that I can think of is to generate a series of SpeedAction with constant rate, but rate increasing or decreasing for each action. 

2. I think your idea with a local static timer is good enough. Here's an example of how it could be defined in the ACC controller:
https://github.com/esmini/esmini/commit/70c342724dd06d9c98ca22de6d4ba7593a65f9fa
There is (yet) no plug-in architecture for controllers. However, they must not be registered in the ScenarioReader module. It can just as well be done from another module, e.g. main.cpp. Here's an example:
https://github.com/esmini/esmini/commit/c7be432442061731fd7745bd5a5fdb5221890e84
The two commits are on branch [controller_timer](https://github.com/esmini/esmini/commits/controller_timer). 
(and I had to remove an obsolete definition in  of ScenarioPlayer in ControllerACC.hpp for this to work. You need to pull either master or the controller_timer branch to get that fix).

Finally: The lateral distance to current lane center is simply pos_.GetOffset()

Good luck!

**Comment by tmIdiada** [comment]
Hi Emil,

could you please look at the following controller, I think it should work as expected (very probably bugs included, because of poor debugging possibilities and rookie knowledge in C++). So, each support from your side would be great and help a lot. Also it would be great if this controller could be integrated into esmini in near future.

[https://unece.org/sites/default/files/2021-03/R157e.pdf](https://unece.org/sites/default/files/2021-03/R157e.pdf) (pages 43-45)


So the logic behind:
- vehicle on adjacent lane with lateral velocity in direction of ego and a deviation from its lane center > 0.375m in direction of ego = cut-in. If the vehicle after cut-in (no full wrap needed, any wrap >= 0) would be within TTC <= 2sec, then ego full braking, if additionally after full wrap the vehicle would be at TTC <= 2sec, then AEB full braking
- vehicle on same lane with deviation from its lane center > 0.375m and a lateral velocity in the same direction means cut-out. After the cut-out was detected and the next vehicle in front of cut-out vehicle is at TTC <= 2sec, then ego full brake. If additionally the cut-out vehicle completely left ego's driving path (full wrap of ego with the second vehicle in front), then AEB full brake
- if the next vehicle in front is braking with >= 5m/s2 and is within TTC <= 2sec, then ego and AEB full brake
- All other cases are not really described, so I expect especially for deceleration maneuvers, that these scenarios are easily to handle and we can use your default ACC controller

What is meaning ego full braking or AEB full braking:
- ego full brake: wait 0.75sec (perception + reaction), then decelerate linearly up to 0.774G within 0.6sec
- AEB full brake: (TTC AEB not found, but everywhere are 2sec taken, so I expect here the same TTC, the delay seems only to be in case of cut-in and cut-out the waiting until the full wrap is there). decelerate directly up to 0.85G within 0.6sec

Please read my comments and my doubts, especially regarding the ordering of the objects in the for loop, because it is really important that the cut-out vehicle in front was at first checked after the next vehicle will be checked inside the for loop.

Kind regards,

Thaddäus
[ECE_ALKS_reference_driver_with_AEB.zip](https://github.com/esmini/esmini/files/7224055/ECE_ALKS_reference_driver_with_AEB.zip)

**Comment by tmIdiada** [comment]
Hi,

after I've integrated that (Controller.hpp + ScenarioReader.cpp) and added the following lines to ego vehicle:
            <ObjectController>
               <Controller name="ECE_ALKS_DRIVER_Controller">
                   <Properties>
                       <Property name="timeGap" value="1.0" />
                   </Properties>
               </Controller>
            </ObjectController>

I'm getting the following error and the controller is definetly not used, due to missing (for debugging I've added many log messages, no other debug possibilities known for esmini) LOG messages:

Scenegraph: None
**Missing esminiController property, using controller name: ECE_ALKS_DRIVER_Controller**
0.000: Log all vehicle data in csv file

Any idea?

Kind regards,

Thaddäus

**Comment by eknabevcc** [comment]
Have you registered it in ScenarioReader.cpp? Add:
`#include "ControllerECE_ALKS_DRIVER.hpp"`
and (among the other similar lines):
`RegisterController(ControllerECE_ALKS_DRIVER::GetTypeNameStatic(), InstantiateControllerECE_ALKS_DRIVER);`

**Comment by eknabevcc** [comment]
Then you should change the loop inside the controller into:
`for (int i = static_cast<int>(entities_->object_.size()) - 1; i >= 0; i--)`
to avoid wrap issue with unsigned type size_t (0 - 1 = some large number)

**Comment by tmIdiada** [comment]
Yes the integration in ScenarioReader is good, what I've forgotten is to add the controller catalogue location to xosc file, is this important, can this be the reason?

**Comment by eknabevcc** [comment]
Good.
I have one additional idea. Maybe the module was not detected by the cmake process. Which looks in the Modules/Controllers directory and grabs all cpp/hpp files in there.
So my advice is to re-run cmake procedure after added the new controller files.

If still problem, I can commit a branch which has your controller integrated and "working".

**Comment by eknabevcc** [comment]
Please check branch [alks](https://github.com/esmini/esmini/tree/alks) and first commit 76acfa3b9de229084a4d53164f087438e28d3bc5

**Comment by tmIdiada** [comment]
Can I also commit and push into this branch?

**Comment by tmIdiada** [comment]
Hi Emil, many thanks for the support

Ok, your example is running and I'm getting my log-outputs from controller :-), my example is not running (still same error message). The single difference I see is the activate controller action, this I've also missed in my xosc file.
Your xosc files like alks-test.xosc are in xosc dir, why is the relative path to the catalogs always **..**/xosc/....????

Kind regards,

Thaddäus

**Comment by eknabevcc** [comment]
> Can I also commit and push into this branch?

I think the best way is to fork the esmini repo. In that fork you'll be the owner and you can commit and push to the branch (you could also create new branches). 

You can then create Pull Request to merge into main repo. It's a little tricky first time, but here is a good description:

https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork

**Comment by eknabevcc** [comment]
> Hi Emil, many thanks for the support
> 
> Ok, the reason was, that I've forgotten to add the controller activate actions to ego vehicle, now it works and I can debug. Your example xosc helped a lot.
> Your xosc files like alks-test.xosc are in xosc dir, why is the relative path to the catalogs always **..**/xosc/....????
> 
> Kind regards,
> 
> Thaddäus

The path "../xosc/..." looks strange I agree. It's an old pattern, I think it was to support to have various scenario-folders like:
```
resources/xosc/
resources/my_scenarios
resources/Thaddäus_scenarios
...
```
And still have a common relative path to the xosc/Catalogs folder. But nowadays path handling is more flexible, e.g. you can add `--path <directory>` to make esmini look in specific directories, so perhaps the ../xosc is not necessary anymore.

**Comment by eknabevcc** [comment]
Regarding GetVelY(). The method returns the y component of the velocity in terms of global coordinate system. I think they are not what you need, since they depends on current heading of vehicle. 

However, I added new methods to retrieve lateral and longitudinal velocity to "our" branch, commit 3443979dffee631c21479a08eb9db20af97ca230

You can try, but please note they are experimental and not really tested yet.

**Comment by tmIdiada** [comment]
> Regarding GetVelY(). The method returns the y component of the velocity in terms of global coordinate system. I think they are not what you need, since they depends on current heading of vehicle.
> 
> However, I added new methods to retrieve lateral and longitudinal velocity to "our" branch, commit [3443979](https://github.com/esmini/esmini/commit/3443979dffee631c21479a08eb9db20af97ca230)
> 
> You can try, but please note they are experimental and not really tested yet.

Hi,

this is great, but it could be, that I need here GetVelT, could you also add GetVelT and GetVelS?
And I've already seen, that my formulas for the jerk are wrong, I'm already working on it, yes I will try it with the fork and pull request. I thought I cannot add a new branch in general due to failing writing permissions in the repository, but if this works, it's totally ok for me by fork.

Kind regards,

Thaddäus

**Comment by eknabevcc** [comment]
The intention of the methods GetVelLat() and GetVelLong() is actually to give velocity relative vehicle heading, so I think it would be the same as GetVelT() and GetVelS().  

But I realized there was a bug. Please get latest commit: 6c12e29b567097eea41e8a02ffd6ee69feac1b72 which should fix it.

**Comment by tmIdiada** [comment]
> The intention of the methods GetVelLat() and GetVelLong() is actually to give velocity relative vehicle heading, so I think it would be the same as GetVelT() and GetVelS().
> 
> But I realized there was a bug. Please get latest commit: [6c12e29](https://github.com/esmini/esmini/commit/6c12e29b567097eea41e8a02ffd6ee69feac1b72) which should fix it.

Hi Emil,

In case of lane change it should be not the same (GetVelLat is the lateral velocity of the vehicle regarding its driving trajectory, GetVelY is the relative velocity regarding the global coordinate frame and finally getVelT would be the relative velocity regarding the driving lane and this is what I would need and this one is defining in the "rate" of the cubic lane change curve, right?).

Kind regards,

Thaddäus

**Comment by eknabevcc** [comment]
Of course, you're right, it should be different. Very good.
I'll look into this later today or tomorrow, and add the GetVelT() and GetVelS() methods.

**Comment by tmIdiada** [comment]
Hi,

I've also seen, that for deceleration maneuver and the 5m/s² condition I would also need a GetAccS() function ;). Hope these things make not too much work.

Many thanks Emil for the support. Meanwhile I've tested my actual driver model at least for cut-in and it works well, except of the exact jerk maybe, the values of GetAccX() and my jerk definition seem not be consistent, maybe you can help me here a bit:

a(t) = -t / 0.6 * 0.774 * 9.81,   t in [0, 0.6], this should be the jerk right?
v(t) is the primitive of a regarding t, so I would expect here v(t) = v0-t**2 / 1.2 * 0.774 * 9.81 for t in [0, 0.6]

Do you see any logical error from my side? I'm getting from GetAccX() values until -25.31, but in general there is something wrong in the usage of this function, this means I'm setting the new velocity values with your SetSpeed from your ACC controller and getting the corresponding acceleration values until 4 time steps later, very strange, like you can see on the messages below.

1.200: Setting speed 5.56 previous speed 5.56
1.200: accX 0.00
1.300: driverFullBrake braking to speed 4.92 within timeStep 0.10
1.300: Setting speed 4.92 previous speed 5.56
1.300: accX 0.00
1.400: driverFullBrake braking to speed 2.39 within timeStep 0.10
1.400: Setting speed 2.39 previous speed 4.92
1.400: accX 0.00
1.500: driverFullBrake braking to speed 0.00 within timeStep 0.10
1.500: Setting speed 0.00 previous speed 2.39
1.500: accX -6.33
1.600: stop 0.00 reset controller
1.600: Setting speed 0.00 previous speed 0.00
1.600: accX -25.31
1.700: Setting speed 0.00 previous speed 0.00
1.700: accX -23.92
1.800: stop 0.00 reset controller
1.800: Setting speed 0.00 previous speed 0.00
1.800: accX 0.00

**Comment by eknabevcc** [comment]
I added some more methods in commit 629120a4ac4806584d79a3739d2617d3195fd802:
```
double GetVelLat();
double GetVelLong();
void GetVelLatLong(double& vlat, double& vlong);
double GetAccLat();
double GetAccLong();
void GetAccLatLong(double& alat, double& along);
double GetVelT();
double GetVelS();
void GetVelTS(double& vt, double& vs);
double GetAccT();
double GetAccS();
void GetAccTS(double& at, double& as);
```
The combi-variants are slightly more efficient since the transformation is done only once.

You can do 'git pull' to get the latest commit.

Regarding delay in acceleration (and speed) values: Yes, motion in esmini is simple kinematics, based on speed and timestep. The direction of motion is often based on actions following a mathematical curve or function (e.g. sinusoidal lane change), and heading (typically) calculated as the tangent to the same function. So motion is not exactly a result of current acceleration and heading.
The actual velocity and acceleration vector are instead calculated as a result of motion. E.g. vel_x = dx / dt, acc_x = vel_x / dt. Hence the delay of one timestep for vel and two for acc.
This might change in future, so that motion is based on accelerations and steering. But so far we focused on creating pure and well defined kinematic motion from the OpenSCENARIO descriptions.

**Comment by tmIdiada** [comment]
Was merged, thanks a lot for all the support
